Data/DataModel/Puzzle.cs

```diff
@@ -43,34 +46,70 @@ public class Puzzle
         public int MinPrerequisiteCount { get; set; } = 0;
 
-        [DataType(DataType.Url)]
-        public string PuzzleUrlString { get; set; }
+        /// <summary>
+        /// All of the content files associated with this puzzle
+        /// </summary>
+        public virtual ICollection<ContentFile> Contents { get; set; }
 
+        /// <summary>
+        /// File for the main puzzle (typically a PDF containing the puzzle)
+        /// </summary>
         [NotMapped]
-        public Uri PuzzleUrl
+        public ContentFile PuzzlePDF
         {
-            get { Uri.TryCreate(PuzzleUrlString, UriKind.RelativeOrAbsolute, out Uri result); return result; }
-            set { PuzzleUrlString = value?.ToString(); }
+            get
+            {
+                var puzzlePDFs = from contentFile in Contents ?? Enumerable.Empty<ContentFile>()
+                                 where contentFile.FileType == ContentFileType.Puzzle
+                                 select contentFile;
+                Debug.Assert(puzzlePDFs.Count() <= 1);
+                return puzzlePDFs.FirstOrDefault();
+            }
         }
 
-        [DataType(DataType.Url)]
-        public string AnswerUrlString { get; set; }
-
+        /// <summary>
+        /// File for the main answer (typically a PDF containing the answer)
+        /// </summary>
         [NotMapped]
-        public Uri AnswerUrl
+        public ContentFile AnswerPDF
         {
-            get { Uri.TryCreate(AnswerUrlString, UriKind.RelativeOrAbsolute, out Uri result); return result; }
-            set { AnswerUrlString = value?.ToString(); }
+            get
+            {
+                var answerPDFs = from contentFile in Contents ?? Enumerable.Empty<ContentFile>()
+                                 where contentFile.FileType == ContentFileType.Answer
+                                 select contentFile;
+                Debug.Assert(answerPDFs.Count() <= 1);
+                return answerPDFs.FirstOrDefault();
+            }
         }
 
-        [DataType(DataType.Url)]
-        public string MaterialsUrlString { get; set; }
 
+        /// <summary>
+        /// Files associated with the puzzle, such as media
+        /// </summary>
         [NotMapped]
-        public Uri MaterialsUrl
+        public IEnumerable<ContentFile> Materials
         {
-            get { Uri.TryCreate(MaterialsUrlString, UriKind.RelativeOrAbsolute, out Uri result); return result; }
-            set { MaterialsUrlString = value?.ToString(); }
+            get
+            {
+                return from contentFile in Contents ?? Enumerable.Empty<ContentFile>()
+                       where contentFile.FileType == ContentFileType.PuzzleMaterial
+                       select contentFile;
+            }
         }
+
+        /// <summary>
+        /// Files unlocked by solving this puzzle, often for metapuzzles
+        /// </summary>
+        [NotMapped]
+        public IEnumerable<ContentFile> SolveTokens
```

Comment:
I think this is a great capability and support it 100% - but since tokens are often just alternate answer words that we build a meta out of, I think we need to be careful about naming. Maybe if all these were postfixed with "File" it would be clear - PuzzleFile, AnswerFile, SupportingFiles, TokenFiles...

Reply:
I want to include "solve" to make the unlocking part clear, but I'm fine with "SolveTokenFiles"